---
title: "seqmagick: sequence manipulation"
author: 
- name: Guangchuang Yu
  email: guangchuangyu@gmail.com
  affiliation: Department of Bioinformatics, School of Basic Medical Sciences, Southern Medical University
date: "`r Sys.Date()`"
bibliography: seqmagick.bib
biblio-style: apalike
output:
  prettydoc::html_pretty:
    toc: true
    highlight: github
    theme: cayman
  pdf_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{seqmagick introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r style, echo=FALSE, results="asis", message=FALSE}
knitr::opts_chunk$set(tidy = FALSE,
		   message = FALSE)
```


```{r echo=FALSE, results="hide", message=FALSE}
library(magrittr)
library(Biostrings)
library(muscle)
library("seqmagick")
```

# Download sequences

## Genbank

```{r}
tmpgb <- tempfile(fileext = '.gb')
tmpfa <- tempfile(fileext = '.fa')
download_genbank(acc='AB115403', format='genbank', outfile=tmpgb)
download_genbank(acc='AB115403', format='fasta', outfile=tmpfa)
readLines(tmpgb)[1:10]
readLines(tmpfa)
```

# File conversion

## fasta and phylip conversion

```{r}
fa_file <- system.file("extdata/HA.fas", package="seqmagick")
## use the small subset to save compilation time of the vignette
fa2 <- tempfile(fileext = '.fa')
fa_read(fa_file) %>% bs_filter('ATGAAAGTAAAA', by='sequence') %>% fa_write(fa2, type='interleaved')


alnfas <- tempfile(fileext = ".fas")
fa_read(fa2) %>% bs_aln(quiet=TRUE) %>% fa_write(alnfas)

## phylip format is only for aligned sequences
tmpphy <- tempfile(fileext = ".phy")
fas2phy(alnfas, tmpphy, type = 'sequential')
readLines(tmpphy)[1:3]
```

`seqmagick` supports both `sequential` and `interleaved` formats, users can specify the format by `type` parameter.


```{r}
phy2fas(tmpphy, alnfas, type = 'interleaved')
readLines(alnfas)[1:4]
```


## interleaved and sequential format conversion


```{r}
readLines(fa2)[1:10]
```

This is a `fasta` file in `interleaved` format. We can convert it to `sequential` format.

```{r}
tmpfas <- tempfile(fileext='.fa')
fa_read(fa2) %>% fa_write(tmpfas, type="sequential")
readLines(tmpfas)[1:4]
```

```{r}
tmpphy2 <- tempfile(fileext = '.phy')
phy_read(tmpphy) %>% phy_write(tmpphy2, type="interleaved")
readLines(tmpphy2)[1:4]
```

# Sequence manipulation


```{r}
bs <- fa_read(fa_file)
bs_filter(bs, 'ATGAAAGTAAAA', by='sequence')

aln <- bs_filter(bs, 'ATGAAAGTAAAA', by='sequence') %>% bs_aln(quiet=TRUE)

aln

bs_consensus(aln)
```



# Bugs/Feature requests

If you have any, [let me know](https://github.com/YuLab-SMU/seqmagick/issues). Thx!

# Session info

Here is the output of `sessionInfo()` on the system on which this document was compiled:
```{r echo=FALSE}
sessionInfo()
```


# References
